09_json/mohitnag/cmd/puppy-server/main_test.go

```diff
@@ -0,0 +1,66 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"testing"
+
+	"github.com/anz-bank/go-course/09_json/mohitnag/pkg/puppy"
+	"github.com/anz-bank/go-course/09_json/mohitnag/pkg/puppy/store"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestMain(t *testing.T) {
+	assert := assert.New(t)
+	var buf bytes.Buffer
+	out = &buf
+	args = []string{"--data", "./../../pkg/puppy/store/testdata/many-puppies.json"}
+	main()
+	expected := "{1 dog white 2}\n{1 dog white 2}\n"
+	actual := buf.String()
+	assert.Equal(expected, actual)
+}
+
+func TestMainError(t *testing.T) {
+	assert := assert.New(t)
+	args = []string{"-d", "./../../pkg/puppy/store/testdata/invalid-puppies.json"}
+	assert.Panics(main)
+}
+
+func TestReadJson(t *testing.T) {
+	t.Run("Valid Json File should pass", func(t *testing.T) {
+		assert := assert.New(t)
+		expect := puppy.Puppy{ID: 1, Breed: "dog", Colour: "white", Value: "2"}
+		actual := puppy.Puppy{}
+		buff := readJSON("./../../pkg/puppy/store/testdata/default-puppy.json")
+		err := json.Unmarshal([]byte(buff), &actual)
+		assert.NoError(err)
+		assert.Equal(expect, actual)
+	})
+	t.Run("Non existing file should fail", func(t *testing.T) {
+		assert := assert.New(t)
+		assert.Panics(func() { readJSON("bad path") })
+	})
+}
+
+func TestUnmarshallingError(t *testing.T) {
+	assert := assert.New(t)
+	var buf bytes.Buffer
+	out = &buf
+	args = []string{"--data", "./../../pkg/puppy/store/testdata/invalid-puppy.json"}
+	assert.Panics(main)
+}
+
+func TestInitialisePuppyStoreErrors(t *testing.T) {
```

Comment:
This test is quite unclear with exactly what constitutes an error and how it is triggering those errors, because it is doing too much. I would favour breaking this up into multiple tests. For example I don't see why this test creates another MapStore and does `initialisePuppyStore` again. The first test fails i think because the store already contains a puppy with ID 1, and then you load a file with two more puppies with ID 1. Rather than being that complex, start with a known state and load a file with 1 puppy with a duplicate ID. The have another test that tests the same failure with a different store failing.

Reply:
Tried to make the test more clear. It tests the duplicate error while initialising puppystore